test: add Ed25519 conformance tests

Sukuna0007Abhi opened the pull request:
## Description
This PR adds conformance tests for Ed25519 signatures to address issue #171.

## Changes
- Added 3 new test vector files in testdata/:
  - `sign1-sign-0007.json`: Ed25519 signing test vector
  - `sign1-verify-0007.json`: Ed25519 verification test vector
  - `sign1-verify-negative-0004.json`: Ed25519 negative verification test
- Updated `conformance_test.go` to support OKP (Octet Key Pair) key type and EdDSA algorithm
- Added `mustBase64ToBytes()` helper function for base64url decoding
- Modified `testSign1()` to use nil rand parameter for deterministic EdDSA signatures

## Test Results
All 21 conformance tests now pass (up from 18):
- 8 signing tests (including new Ed25519)
- 8 verification tests (including new Ed25519)
- 5 negative verification tests (including new Ed25519)

## Technical Notes
- Ed25519 signatures are deterministic and require `nil` for the rand parameter, unlike ECDSA/RSA which use `zeroSource`
- Test vectors are based on COSE working group examples from RFC 8152
- All existing tests continue to pass

Fixes #171 

Ready for review sir @thomas-fossati @setrofim @yogeshbdeshpande @SimonFrost-Arm @OR13 @henkbirkholz

Sukuna0007Abhi committed:
Add Ed25519 conformance tests

- Add sign1-sign-0007.json: Ed25519 signing test vector
- Add sign1-verify-0007.json: Ed25519 verification test vector
- Add sign1-verify-negative-0004.json: Ed25519 negative verification test
- Update conformance_test.go to support OKP key type and EdDSA algorithm
- Add mustBase64ToBytes() helper for base64url decoding
- Modify testSign1() to use nil rand for deterministic EdDSA signatures

Fixes #171

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

OR13 commented on conformance_test.go:
```diff
@@ -320,6 +408,8 @@ func mustNameToAlg(name string) cose.Algorithm {
 		return cose.AlgorithmES384
 	case "ES512":
 		return cose.AlgorithmES512
+	case "EdDSA":
```
Technically this could also mean Ed448.

Consider if anything here is helpful:

https://datatracker.ietf.org/doc/draft-ietf-jose-fully-specified-algorithms/13/

Sukuna0007Abhi replied:
Sure sir @OR13 , thanks for the approval

Sukuna0007Abhi replied:
Hi sir @OR13 Good point about Ed448. I've added placeholder support for it with proper error messages and comments about the key size differences (57-byte vs 32-byte seeds). The code structure now makes it easy to add Ed448 implementation later when golang.org/x/crypto/ed448 is integrated. Both curves correctly use AlgorithmEdDSA = -8 with curve differentiation in the key's "crv" field.

Sukuna0007Abhi committed:
Add Ed448 placeholder support and future-proofing comments

- Add Ed448 case with clear 'not yet implemented' message
- Add comments about Ed448 key size differences (57 vs 32 bytes)
- Document that Ed448 would require golang.org/x/crypto/ed448
- Maintain backward compatibility while preparing for future expansion

This addresses the mentor's suggestion about supporting both Ed25519 and Ed448
under the EdDSA algorithm identifier.

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi commented:
Ready for review and merge as sir @OR13 approved it

SteveLasker commented:
@shizhMSFT, @qmuntal can either of you PTAL?

shizhMSFT reviewed (approved):
Thanks for your contribution! No special concerns and LGTM.

I've changed the PR title to [conventional commits](https://www.conventionalcommits.org/) as this PR will be squash merged.

Please also take a look at the comments provided by Copilot and validate if they are relevant.

Sukuna0007Abhi commented:
Thanks sir @shizhMSFT for the approval, I will review the suggestions of copilot and if needed I will update the code,
& No issue with changes of PR title..

yogeshbdeshpande commented on conformance_test.go:
```diff
@@ -248,6 +309,32 @@ func getKey(key Key, private bool) (crypto.Signer, error) {
 			pkey.D = mustBase64ToBigInt(key["d"])
 		}
 		return pkey, nil
+	case "OKP":
+		// Only Ed25519 is supported for now
+		switch key["crv"] {
+		case "Ed25519":
+			publicKey := mustBase64ToBytes(key["x"])
+			if len(publicKey) != ed25519.PublicKeySize {
+				return nil, errors.New("invalid Ed25519 public key size")
+			}
+			if private {
+				privateKey := mustBase64ToBytes(key["d"])
+				if len(privateKey) != ed25519.SeedSize {
+					return nil, errors.New("invalid Ed25519 private key size")
+				}
+				// Ed25519 private key is 64 bytes: 32-byte seed + 32-byte public key
+				fullPrivateKey := ed25519.NewKeyFromSeed(privateKey)
+				return fullPrivateKey, nil
+			}
+			// For public key only operations, we need to return a type that satisfies crypto.Signer
+			// but this won't work for verify-only operations. We'll handle this in getSigner.
+			return nil, errors.New("OKP public-only key not supported in this context")
```
This is a very valid comment, which needs to be incorporated!

shukla7237ashutsosh replied:
Sure sir @yogeshbdeshpande will investigate and focus on it

Sukuna0007Abhi replied:
Ok sir @yogeshbdeshpande I will more review this one...

Sukuna0007Abhi commented on conformance_test.go:
```diff
@@ -153,6 +158,7 @@ func testSign1(t *testing.T, tc *TestCase, deterministic bool) {
 	if err != nil {
 		t.Fatal(err)
 	}
+	t.Logf("Test case: %s, Key type: %s", tc.Title, tc.Key["kty"])
```
This one is great, changes will be made fr...

Sukuna0007Abhi committed:
Address Copilot code review suggestions

- Remove verbose debug logging to match existing test style
- Extract Ed25519 key creation logic into createEd25519Key() helper function
- Eliminate code duplication between getSigner() and getKey() functions
- Improve code maintainability while preserving all functionality

Addresses feedback from @yogeshbdeshpande and GitHub Copilot code review.

Signed-off-by: Sukuna0007Abhi <appsonly310@gmail.com>

Sukuna0007Abhi commented:
Hi sir @shizhMSFT @SteveLasker @yogeshbdeshpande @henkbirkholz @OR13 @thomas-fossati @qmuntal ready for review and merge, as I addressed all copilot suggestions also

shizhMSFT reviewed (approved):
LGTM

Sukuna0007Abhi commented:
Thanks sir @shizhMSFT

yogeshbdeshpande commented on conformance_test.go:
```diff
@@ -187,11 +198,44 @@ func testSign1(t *testing.T, tc *TestCase, deterministic bool) {
+	if tc.Key["kty"] == "OKP" {
+		switch tc.Key["crv"] {
+		case "Ed25519":
+			// Note: Ed448 would require different key size validation (57 bytes)
```
Is comment on line 207 correct ? 
Should it not be Ed25519 ?

Sukuna0007Abhi replied:
Yes sir @yogeshbdeshpande  you're right I didn't put another line though, 

I am actually thought of putting "We're currently handling Ed25519, but if we added Ed448, it would require different key size validation (57 bytes)" 

But now I think it's merged so any member can change and commit this comment change part ,

Thanks great catching sir...

yogeshbdeshpande replied:
@Sukuna0007Abhi , please create a small minor edit PR and work on improving this...

Sukuna0007Abhi replied:
Sure sir @yogeshbdeshpande will do that

SteveLasker reviewed (approved):
Thank you @Sukuna0007Abhi for the contribution and thanks @OR13, @shizhMSFT, @yogeshbdeshpande for the review.
LGTM

Sukuna0007Abhi commented:
Thank you sir @SteveLasker @OR13 @yogeshbdeshpande @shizhMSFT @henkbirkholz @thomas-fossati @setrofim  ;
First contribution on go-cose is successful ❤️‍🔥
